presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineSemiJoinDistributionType.java

```diff
@@ -14,50 +14,86 @@
 package com.facebook.presto.sql.planner.optimizations;
 
 import com.facebook.presto.Session;
+import com.facebook.presto.cost.CachingStatsProvider;
+import com.facebook.presto.cost.CostComparator;
+import com.facebook.presto.cost.PlanNodeCostEstimate;
+import com.facebook.presto.cost.StatsCalculator;
+import com.facebook.presto.cost.StatsProvider;
+import com.facebook.presto.cost.TaskCountEstimator;
 import com.facebook.presto.execution.warnings.WarningCollector;
 import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
 import com.facebook.presto.sql.planner.SymbolAllocator;
 import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.rule.PlanNodeWithCost;
 import com.facebook.presto.sql.planner.plan.DeleteNode;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 import com.facebook.presto.sql.planner.plan.SemiJoinNode;
 import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.collect.Ordering;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Optional;
 
 import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.cost.CostCalculatorWithEstimatedExchanges.calculateJoinExchangeCost;
+import static com.facebook.presto.cost.CostCalculatorWithEstimatedExchanges.calculateJoinInputCost;
+import static com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType.AUTOMATIC;
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.REPLICATED;
 import static java.util.Objects.requireNonNull;
 
 public class DetermineSemiJoinDistributionType
         implements PlanOptimizer
 {
+    private final StatsCalculator statsCalculator;
```

Comment:
I would split this rule into 2 rules. First will be the legacy rewrite that checks if the semi join needs to be `REPLICATED` because of `DELETE`. The second rule should be an iterative rule, similar to `DetermineJoinDistributionType`, that should be applied after the first rule that checks `DELETE`. So then you don't have to inject `StatsCalculator` here, and can reuse `StatsCalculator` from the `Context` of the iterative rule.

Reply:
Once we have better matcher we can migrate the first legacy rule to be an iterative rule as well.